### PR TITLE
Add Trie, JsonObject, and Document

### PR DESCRIPTION
**Pull request by hyeongsik-won**
**What this PR does / why we need it**:

Add Trie, JsonObject, and Document
JsonArray and Document are not fully developed.

In order to proxy the custom data types in JsonObject, dynamicMemeberLookup is used. 
It can be changed to any other type that doesn't use dynamicMemeberLookup.
(ex: get(key: "key"), set(key: "key", value: "value")

Trie and related test codes are based on JS-SDK. Swift doesn't have a useful trie library.

There is no code that requires Observables and Observers in ```Document```. It will be added soon.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation**:

```docs

```

**Checklist**:
- [ ] Added relevant tests or not required
- [ ] Didn't break anything

### CODE REVIEW

File: Sources/Document/Document.swift [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+     *
+     */
+    func hasLocalChanges() -> Bool {
+        return self.localChanges.isEmpty == false
+    }
+
+    /**
+     * `ensureClone` make a clone of root.
+     */
+    func cloned() -> CRDTRoot {
```
Comment (hyeongsik-won) [inline Sources/Document/Document.swift:76]: Changed to return CRDTRoot to ensure that clone t is not nil.
```ensureClone``` is renamed to ```cloned```.

File: Sources/Document/Document.swift [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+     * changes the document has.
+     *
+     */
+    func setActor(_ actorID: ActorID) {
+        for change in self.localChanges {
+            change.setActor(actorID)
+        }
+        self.changeID.setActor(actorID)
+
+        // TODOs also apply into root.
```
Comment (hyeongsik-won) [inline Sources/Document/Document.swift:108]: What is the mean ```// TODO also apply into root.```?

File: Sources/Document/Document.swift [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+     */
+    func getRootObject() -> CRDTObject {
+        return self.root.getObject()
+    }
+
+    /**
+     * `getGarbageLength` returns the length of elements should be purged.
+     *
+     */
+    func getGarbageLength() -> Int {
```
Comment (hyeongsik-won) [inline Sources/Document/Document.swift:161]: ```getGarbageLen``` is renamed to ```getGarbageLength```.

File: Sources/Document/Json/JsonArray.swift [hunk 4]
```diff
@@ -15,3 +15,23 @@
  */
 
 import Foundation
+
+// It will be implemented soon.
+protocol JsonArrayable: AnyObject {
```
Comment (humdrum) [inline Sources/Document/Json/JsonArray.swift:20]: How about changing "Json"  to "JSON" like HTTP

File: Sources/Document/Document.swift (hunk 2 above)
Comment (hackerwins) [reply Sources/Document/Document.swift:108]: The issue below is detailed about this.
https://github.com/yorkie-team/yorkie/issues/417

**Review by hackerwins — APPROVED** [review]
Thanks for your contribution. 👍

**Review by humdrum — APPROVED** [review]
LGTM!

**Pull request merged**